### PR TITLE
make: add submodule support

### PR DESCRIPTION
**Pull request by kaspar030**
This PR makes it possible for modules to not always use e.g., ```*.c``` as source files, but conditionally select some files as so-called "submodule".

See the adapted core/ Makefile as example. Previously, ugly ```#ifdef MODULE_CORE_MSG``` in msg.c controlled compiling that functionality. Now ```USEMODULE += core_msg``` does.

### CODE REVIEW

**Comment by lebrush** [comment]
Could this be adaptable to `conn_*`, `gnrc_*`, etc?

**Comment by kaspar030** [comment]
> Could this be adaptable to conn_*, gnrc_*, etc?

Sure, this is usable wherever "submodules" consist of a single *.c file. gnrc_sock_* would be a low hanging fruit.

I actually devised this for periph_*, but the way out FEATURES work is blocking the way...

File: Makefile.base [hunk 1]
```diff
@@ -22,16 +22,33 @@ ${DIRS:%=ALL--%}:
 ${DIRS:%=CLEAN--%}:
 	$(QQ)"$(MAKE)" -C ${@:CLEAN--%=%} clean
 
-ifeq ($(strip $(SRC)),)
+## submodules
+ifeq (1, $(SUBMODULES))
+    # don't use *.c as SRC if SRC is empty (e.g., no module selected)
+    NO_AUTO_SRC := 1
+
+    # allow different submodule basename (e.g., MODULE=cpu_periph_common, but match just periph_%
```
Comment (haukepetersen) [inline Makefile.base]: )

**Comment by haukepetersen** [comment]
Except this one parenthesis: I love it!

**Comment by kaspar030** [comment]
- addressed and squashed

**Pull request merged**